### PR TITLE
Don't allocate a stack

### PR DESCRIPTION
**Pull request by davidfowl**
- Avoid an extra object allocation by using a struct based stack

### CODE REVIEW

File: src/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegmentStack.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
+﻿using System;
```
Comment (jkotalik) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/BufferSegmentStack.cs]: copyright.

**Comment by benaadams** [comment]
If you are going as far as adding custom Stack, might as well make the array non-covariant by using a [`RefAsValueType`](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/RefAsValueType.cs) like type? Though implicit conversions are the way to go as it reduces code changes:

```csharp
public partial struct BufferSegmentStack
{
    // Change from BufferSegment[] to SegmentAsValueType[]
    private SegmentAsValueType[] _array;

    // Add
    private readonly struct SegmentAsValueType
    {
        private readonly BufferSegment _value;
        private SegmentAsValueType(BufferSegment value) => _value = value;
        public static implicit operator SegmentAsValueType(BufferSegment s) => new SegmentAsValueType(s);
        public static implicit operator BufferSegment(SegmentAsValueType s) => s._value;
    }
}
```

**Comment by benaadams** [comment]
And maybe add a comment for it as a variation on 
```csharp
    /// <summary>
    /// A simple struct we wrap reference types inside when storing in arrays to
    /// bypass the CLR's covariant checks when writing to arrays.
    /// </summary>
    /// <remarks>
    /// We use <see cref="RefAsValueType{T}"/> as a wrapper to avoid paying the cost of covariant checks whenever
    /// the underlying array that the <see cref="Stack{T}"/> class uses is written to. 
    /// We've recognized this as a perf win in ETL traces for these stack frames:
    /// clr!JIT_Stelem_Ref
    ///   clr!ArrayStoreCheck
    ///     clr!ObjIsInstanceOf
```

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 2]
```diff
@@ -44,7 +44,7 @@ public sealed partial class Pipe
         private readonly PipeScheduler _readerScheduler;
         private readonly PipeScheduler _writerScheduler;
 
-        private readonly Stack<BufferSegment> _bufferSegmentPool;
+        private readonly BufferSegmentStack _bufferSegmentPool;
```
Comment (pakrym) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: How does this work? I thought readonly struct fields are copied before method invocations.

Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Ah frack, I thought I changed this. I'll add a comment too.

Comment (pakrym) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: I'd rather you added a test.

**Pull request merged**